### ISSUE TITLE
When using a TAI or SPNEGO, provide a way to prevent the server from returning LTPA cookies

### ISSUE THREAD

**Issue by jtmulvey** [OpenLiberty/open-liberty#9445]
We've had several requests from customers that we not blindly create and return LTPA (Liteweight Third-Party Access) tokens/cookies based on the presence of a Trust Association Interceptor (TAI) or when using our SPNEGO support.  With this, we intend to introduce a new option to allow customers to keep us from unconditionally doing this.

WAD: 
https://ibm.box.com/s/0yhz0vfbzd0gzt9g35jh6t1uc8c4os6t

**Comment by atosak** [comment]
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - -a- - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [ ] POC Design / WAD Review Scheduled (David Chang) or N/A.
- [ ] POC Design / WAD Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [ ] Design / WAD Approval (Alasdair Nottingham) or N/A.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of the epic entitled "FAT Approval Test Summary".  Add and fill in the template as described here: https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [ ] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approved or No Design Approved label on the GitHub Epic.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Sam Wong - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
- [ ] **Demo** - (Liberty only - Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by garypicher** [comment]
Design review follow-on actions:

- interested users slide, fix spelling of "SwissRe" and add "Svenska Handelsbanken"
- feature design slide, should we do something similar with JWT cookies as part of this feature, or as a separate feature?  set up a conversation with alasdair to agree on a plan
- admin / config / command line slide, should there be a global option to disable LTPA cookie no matter *what* authentication mechanism is being used, as opposed to a separate option for each config element?  set up a conversation with alasdair to agree on a plan
- based on the above conversations, alasdair will determine whether a 2nd socialization is needed

**Comment by gscottj** [comment]
No AVT required. This feature does not change the UI or add any new UI. This feature adds a new API to provide a way to programmatically retrieve the user name from the LTPA token.

**Comment by utle** [comment]
DEV: https://github.com/OpenLiberty/open-liberty/issues/11669
FAT: https://github.com/OpenLiberty/open-liberty/issues/11670

**Comment by utle** [comment]
Doc issue:  https://github.com/OpenLiberty/docs/issues/1554

**Comment by utle** [comment]
https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_sec_custmr.html
Add the following before the Procedure: 
custom TAI as a Liberty feature support properties invokeBeforeSSO and invokeAfterSSO similar to the custom TAI as a shared library configuration (link to the metatype??) .  If you do not specify these two properties, the custom TAI will be called after SSO.

You can specify these properties using the annotation @Component. For example:
`@Component(service = { TrustAssociationInterceptor.class }, immediate = true, configurationPolicy = ConfigurationPolicy.IGNORE, name = "sampleTAI", property = { "service.vendor=IBM", "type=sampleTAI", "id=sampleTAI", "TAIName=sampleTAI", **"invokeBeforeSSO:Boolean=true"}**)`

You can also specify these properties in the OSGI-INF manifest for src:component as a property. For example
```
<?xml version="1.0" encoding="UTF-8"?>
<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.3.0" name="sampleTAI" configuration-policy="ignore" immediate="true" activate="activate" deactivate="deactivate" modified="modified">
  <property name="service.vendor" type="String" value="IBM"/>
  <property name="type" type="String" value="sampleTAI"/>
  <property name="id" type="String" value="sampleTAI"/>
  <property name="TAIName" type="String" value="sampleTAI"/>
  <property name="invokeBeforeSSO" type="Boolean" value="true"/>
 
  <service>
    <provide interface="com.ibm.wsspi.security.tai.TrustAssociationInterceptor"/>
  </service>
 
  <implementation class="com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI"/>
</scr:component>
```

**Comment by Zech-Hein** [comment]
Feature Test Summary: #12512

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by utle** [comment]
Hi @donbourne, I have updated the Serviceability Approval. Thanks.

**Comment by donbourne** [comment]
Hi Ut, for question 3 it usually works best if you can find one or two folks to demo the problem flows to.  That's not something that is usually done at the end of iteration demo.  The people you demo to can take a minute to put themselves in customers' shoes and decide if they think the messages or other diagnostics you provide are sufficient to solve the problems they might encounter.  Can you please do a demo like that?  Should be pretty quick if there aren't many problem paths to consider.

**Comment by utle** [comment]
Hi @donbourne, I updated question above. Thank you.

**Comment by atosak** [comment]
@utle , I unchecked the STE in the checklist as Swati confirmed no STE have been received yet and therefore no approval.

**Comment by utle** [comment]
Hi @atosak L2 team agreed there is no STE needed. I will forward you the email.

**Comment by chirp1** [comment]
Approving. As I mentiond in a slack to Ut, the issue is on the doc backlog and will be prioritized with the other issues. Likely, the function will eGA before the doc is written. So, have enough info/examples in the eGA blog post so that a user can use the function before the final topic is written and published.

**Comment by donbourne** [comment]
> Hi @donbourne, I updated question 3 above. Thank you.

I actually meant question 2 -- but I think you've answered what I need in your update to question 3 :-)

**Comment by krismarc** [comment]
@utle, I'd appreciate if you could take a look at https://github.com/OpenLiberty/open-liberty/issues/32360
..not sure if I am doing anything wrong, I've just tried adding attribute to spnego tag within server.xml
